# Data Import and Cleaning 

## Part 1 of 2: ACT 2019 Dataset for California

*All libraries used should be added here*

In [1]:
import pandas as pd 

## Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind the minimum & maximum possible values for each test/subtest).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of percents (ex. '50%', '30.5%', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to floats! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

## Loading the data 

### Reading in the .csv files

In [2]:
act = pd.read_csv('../data/act_2019_ca.csv')

In [3]:
act.head()

CDS  CCode     CDCode      SCode RType  \
0  3.366993e+13   33.0  3366993.0   129882.0     S   
1  1.964212e+13   19.0  1964212.0  1995596.0     S   
2  1.563776e+13   15.0  1563776.0  1530377.0     S   
3  4.369666e+13   43.0  4369666.0  4333795.0     S   
4  1.964733e+13   19.0  1964733.0  1935121.0     S   

                             SName                  DName        CName  \
0  21st Century Learning Institute       Beaumont Unified    Riverside   
1      ABC Secondary (Alternative)            ABC Unified  Los Angeles   
2      Abraham Lincoln Alternative  Southern Kern Unified         Kern   
3             Abraham Lincoln High       San Jose Unified  Santa Clara   
4      Abraham Lincoln Senior High    Los Angeles Unified  Los Angeles   

   Enroll12  NumTstTakr AvgScrRead AvgScrEng AvgScrMath AvgScrSci NumGE21  \
0      18.0         0.0        NaN       NaN        NaN       NaN     NaN   
1      58.0         0.0        NaN       NaN        NaN       NaN     NaN   
2      18.0         0.0        NaN       NaN        NaN       NaN     NaN   
3     463.0        53.0         23        22         22        23      34   
4     226.0        19.0         21        20         23        22      11   

  PctGE21     Year  Unnamed: 17  
0     NaN  2018-19          NaN  
1     NaN  2018-19          NaN  
2     NaN  2018-19          NaN  
3   64.15  2018-19          NaN  
4   57.89  2018-19          NaN

### Shape of the data 

The dataset has the following number of (rows, columns) respectively. 

In [4]:
act.shape

(2310, 18)

### Removing the extra column at the end

In [5]:
act.drop('Unnamed: 17', axis=1, inplace=True)

In [6]:
act.head(3)

CDS  CCode     CDCode      SCode RType  \
0  3.366993e+13   33.0  3366993.0   129882.0     S   
1  1.964212e+13   19.0  1964212.0  1995596.0     S   
2  1.563776e+13   15.0  1563776.0  1530377.0     S   

                             SName                  DName        CName  \
0  21st Century Learning Institute       Beaumont Unified    Riverside   
1      ABC Secondary (Alternative)            ABC Unified  Los Angeles   
2      Abraham Lincoln Alternative  Southern Kern Unified         Kern   

   Enroll12  NumTstTakr AvgScrRead AvgScrEng AvgScrMath AvgScrSci NumGE21  \
0      18.0         0.0        NaN       NaN        NaN       NaN     NaN   
1      58.0         0.0        NaN       NaN        NaN       NaN     NaN   
2      18.0         0.0        NaN       NaN        NaN       NaN     NaN   

  PctGE21     Year  
0     NaN  2018-19  
1     NaN  2018-19  
2     NaN  2018-19

The column has been removed.

### Standardising the format of the column names 

The column names will be changed to lowercase, and with underscores as word separators.

In [7]:
# Changing column names to lowercase
act.rename(str.lower, axis="columns", inplace=True)

In [8]:
act.head(1)

cds  ccode     cdcode     scode rtype  \
0  3.366993e+13   33.0  3366993.0  129882.0     S   

                             sname             dname      cname  enroll12  \
0  21st Century Learning Institute  Beaumont Unified  Riverside      18.0   

   numtsttakr avgscrread avgscreng avgscrmath avgscrsci numge21 pctge21  \
0         0.0        NaN       NaN        NaN       NaN     NaN     NaN   

      year  
0  2018-19

In [9]:
# Renaming the column names with underscores as word separators
act.rename(columns = {'cds':'cds_code', 
                      'ccode':'c_code',
                      'cdcode':'cd_code',
                      'scode':'s_code',
                      'rtype':'r_type',
                      'sname':'s_name',
                      'dname':'d_name',
                      'cname':'c_name',
                      'enroll12':'enroll_12',
                      'numtsttakr':'num_tst_takr',
                      'avgscrread':'avg_scr_read',
                      'avgscreng':'avg_scr_eng',
                      'avgscrmath':'avg_scr_math',
                      'avgscrsci':'avg_scr_sci',
                      'numge21':'num_ge_21',
                      'pctge21':'pct_ge_21' 
                     }, inplace=True)

In [10]:
act.head(1)

cds_code  c_code    cd_code    s_code r_type  \
0  3.366993e+13    33.0  3366993.0  129882.0      S   

                            s_name            d_name     c_name  enroll_12  \
0  21st Century Learning Institute  Beaumont Unified  Riverside       18.0   

   num_tst_takr avg_scr_read avg_scr_eng avg_scr_math avg_scr_sci num_ge_21  \
0           0.0          NaN         NaN          NaN         NaN       NaN   

  pct_ge_21     year  
0       NaN  2018-19

The columns have been renamed.

### Checking for null values

In [11]:
act.isna().sum()

cds_code          1
c_code            1
cd_code           1
s_code          523
r_type            1
s_name          581
d_name           59
c_name            1
enroll_12         1
num_tst_takr      1
avg_scr_read    357
avg_scr_eng     357
avg_scr_math    357
avg_scr_sci     357
num_ge_21       357
pct_ge_21       357
year              1
dtype: int64

Let's look closer into where these null values appear. 

In [12]:
# Fetch the rows of the 'act' DataFrame where there is a null value under the 'cds_code' column.
act[act['cds_code'].isna()]

cds_code  c_code  cd_code  s_code r_type s_name d_name c_name  \
2309       NaN     NaN      NaN     NaN    NaN    NaN    NaN    NaN   

      enroll_12  num_tst_takr avg_scr_read avg_scr_eng avg_scr_math  \
2309        NaN           NaN          NaN         NaN          NaN   

     avg_scr_sci num_ge_21 pct_ge_21 year  
2309         NaN       NaN       NaN  NaN

This row seems to be the last row, according to the shape of the data that we retrieved just now. Let's check:

In [13]:
act.tail(3)

cds_code  c_code    cd_code     s_code r_type             s_name  \
2307  3.667777e+13    36.0  3667777.0  3638616.0      S  Yucca Valley High   
2308  3.667850e+13    36.0  3667850.0  3630530.0      S       Zupanic High   
2309           NaN     NaN        NaN        NaN    NaN                NaN   

               d_name          c_name  enroll_12  num_tst_takr avg_scr_read  \
2307  Morongo Unified  San Bernardino      314.0          11.0            *   
2308   Rialto Unified  San Bernardino       47.0           0.0          NaN   
2309              NaN             NaN        NaN           NaN          NaN   

     avg_scr_eng avg_scr_math avg_scr_sci num_ge_21 pct_ge_21     year  
2307           *            *           *         *         *  2018-19  
2308         NaN          NaN         NaN       NaN       NaN  2018-19  
2309         NaN          NaN         NaN       NaN       NaN      NaN

### Dropping the last row

It seems like the last row is entirely made up of null values. We will drop the row. 

In [14]:
act.drop(2309, inplace=True)

In [15]:
act.isna().sum()

cds_code          0
c_code            0
cd_code           0
s_code          522
r_type            0
s_name          580
d_name           58
c_name            0
enroll_12         0
num_tst_takr      0
avg_scr_read    356
avg_scr_eng     356
avg_scr_math    356
avg_scr_sci     356
num_ge_21       356
pct_ge_21       356
year              0
dtype: int64

We have removed null values for many of the columns. 

### Checking the `'s_code'` column

Now, let's look at the next column with null values, the `'s_code'` column. We know from the data source that this stands for the School Code. (We also know what the rest of the column names mean.)

In [16]:
# Fetch the first 3 rows of the 'act' DataFrame, where there is a null value under the 's_code' column.
act[act['s_code'].isna()].head(3)

cds_code  c_code   cd_code  s_code r_type s_name  \
1122  1.100170e+12     1.0  110017.0     NaN      D    NaN   
1123  1.611190e+12     1.0  161119.0     NaN      D    NaN   
1124  1.611270e+12     1.0  161127.0     NaN      D    NaN   

                                  d_name   c_name  enroll_12  num_tst_takr  \
1122  Alameda County Office of Education  Alameda      170.0           0.0   
1123                     Alameda Unified  Alameda      919.0         155.0   
1124                 Albany City Unified  Alameda      307.0          58.0   

     avg_scr_read avg_scr_eng avg_scr_math avg_scr_sci num_ge_21 pct_ge_21  \
1122          NaN         NaN          NaN         NaN       NaN       NaN   
1123           27          26           26          25       134     86.45   
1124           28          27           27          26        51     87.93   

         year  
1122  2018-19  
1123  2018-19  
1124  2018-19

The entries under the `'r_type'` (Record Type) are labelled as 'D' (District). Let's see if the entire column is labelled as such. 

In [17]:
# Fetch the value counts of the 'r_type' column of the 'act' DataFrame, where there is a null value under the 'SCode' column.
act[act['s_code'].isna()]['r_type'].value_counts()

D    522
Name: r_type, dtype: int64

It seems like all the entries with NaN school codes are of the Record Type 'District'. Hence, they are aggregated data of a number of schools. It makes sense that the School Codes are NaN. 

### Replacing the null values in the `'s_code'` column 

Let's first see how the other (non-null) values look like:

In [18]:
act.head(3)

cds_code  c_code    cd_code     s_code r_type  \
0  3.366993e+13    33.0  3366993.0   129882.0      S   
1  1.964212e+13    19.0  1964212.0  1995596.0      S   
2  1.563776e+13    15.0  1563776.0  1530377.0      S   

                            s_name                 d_name       c_name  \
0  21st Century Learning Institute       Beaumont Unified    Riverside   
1      ABC Secondary (Alternative)            ABC Unified  Los Angeles   
2      Abraham Lincoln Alternative  Southern Kern Unified         Kern   

   enroll_12  num_tst_takr avg_scr_read avg_scr_eng avg_scr_math avg_scr_sci  \
0       18.0           0.0          NaN         NaN          NaN         NaN   
1       58.0           0.0          NaN         NaN          NaN         NaN   
2       18.0           0.0          NaN         NaN          NaN         NaN   

  num_ge_21 pct_ge_21     year  
0       NaN       NaN  2018-19  
1       NaN       NaN  2018-19  
2       NaN       NaN  2018-19

The rest of the values seem to be formatted as a float. We will thus swap out the NaN values for the float 0.0. 

In [19]:
act['s_code'].fillna(value=0.0, inplace=True)

In [20]:
# Fetch the value counts of the 's_code' column of the 'act' DataFrame, where the value under the'r_type' column is equal to 'D'.
act[act['r_type'] == 'D']['s_code'].value_counts()

0.0    522
Name: s_code, dtype: int64

We can see that the intended values have been replaced, and the value is still formatted as 0.0. 

In [21]:
act.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2309 entries, 0 to 2308
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cds_code      2309 non-null   float64
 1   c_code        2309 non-null   float64
 2   cd_code       2309 non-null   float64
 3   s_code        2309 non-null   float64
 4   r_type        2309 non-null   object 
 5   s_name        1729 non-null   object 
 6   d_name        2251 non-null   object 
 7   c_name        2309 non-null   object 
 8   enroll_12     2309 non-null   float64
 9   num_tst_takr  2309 non-null   float64
 10  avg_scr_read  1953 non-null   object 
 11  avg_scr_eng   1953 non-null   object 
 12  avg_scr_math  1953 non-null   object 
 13  avg_scr_sci   1953 non-null   object 
 14  num_ge_21     1953 non-null   object 
 15  pct_ge_21     1953 non-null   object 
 16  year          2309 non-null   object 
dtypes: float64(6), object(11)
memory usage: 324.7+ KB


The datatype is a float. 

### Checking the `'s_name'` column 

This is the next column with null values in the list, for School Names.

In [22]:
# Fetch the first 3 rows of the 'act' DataFrame, where there is a null value under the 's_name' column.
act[act['s_name'].isna()].head(3)

cds_code  c_code  cd_code  s_code r_type s_name d_name       c_name  \
1064  3.400000e+13    34.0      0.0     0.0      C    NaN    NaN   Sacramento   
1065  3.900000e+13    39.0      0.0     0.0      C    NaN    NaN  San Joaquin   
1066  1.500000e+13    15.0      0.0     0.0      C    NaN    NaN         Kern   

      enroll_12  num_tst_takr avg_scr_read avg_scr_eng avg_scr_math  \
1064    19540.0        3065.0           21          20           21   
1065    11778.0        1348.0           22          21           21   
1066    14229.0        1257.0           21          20           20   

     avg_scr_sci num_ge_21 pct_ge_21     year  
1064          21      1526     49.79  2018-19  
1065          21       693     51.41  2018-19  
1066          20       544     43.28  2018-19

In [23]:
# Fetch the last 3 rows of the 'act' DataFrame, where there is a null value under the 's_name' column.
act[act['s_name'].isna()].tail(3)

cds_code  c_code    cd_code  s_code r_type s_name  \
1641  5.872728e+13    58.0  5872728.0     0.0      D    NaN   
1642  5.872736e+13    58.0  5872736.0     0.0      D    NaN   
1643  5.872769e+13    58.0  5872769.0     0.0      D    NaN   

                        d_name c_name  enroll_12  num_tst_takr avg_scr_read  \
1641   Camptonville Elementary   Yuba       42.0           4.0            *   
1642  Marysville Joint Unified   Yuba      602.0          42.0           20   
1643      Wheatland Union High   Yuba      167.0          18.0           23   

     avg_scr_eng avg_scr_math avg_scr_sci num_ge_21 pct_ge_21     year  
1641           *            *           *         *         *  2018-19  
1642          18           19          19        17     40.48  2018-19  
1643          21           21          22        10     55.56  2018-19

Here, we can see that the entries with no school name only have a `'c_name'` (County Name), or have both a `'d_name'` (District Name) and a `'c_name'` (County Name). Hence, it makes sense that these rows should be for the aggregated values of districts and counties, respectively. 

In [24]:
# Fetch the value counts of the 'r_type' column where the values of the 's_name' column are null
act[act['s_name'].isna()].r_type.value_counts()

D    522
C     58
Name: r_type, dtype: int64

Using the `'r_type'` (Record Type) column, there are indeed only record types for districts and counties under entries with no school names. Do these represent all the district and county records in the DataFrame?

In [25]:
act['r_type'].value_counts()

S    1728
D     522
C      58
X       1
Name: r_type, dtype: int64

Yes, they do. However, we see an 'X' value. According to the data source, this is actually the state-level record (for all of California).

In [26]:
# Fetch the rows of the 'act' DataFrame, where the value of the 'r_type' column is 'X'.
act[act['r_type'] == 'X']

cds_code  c_code  cd_code  s_code r_type               s_name  \
2070       0.0     0.0      0.0     0.0      X  State of California   

                   d_name               c_name  enroll_12  num_tst_takr  \
2070  State of California  State of California   489650.0       82668.0   

     avg_scr_read avg_scr_eng avg_scr_math avg_scr_sci num_ge_21 pct_ge_21  \
2070           22          22           22          22     45466     55.00   

         year  
2070  2018-19

There is no column for state name, which is most likely the reason for entering it under `'s_name'` (School Name). We can take note of this. 

### Replacing the null values in the `'s_name'` column 

We will simply fill the null values with the string 'None'.

In [27]:
act['s_name'].fillna(value='None', inplace=True)

### Checking the `'d_name'` column 

These are the District Names. 

In [28]:
# Fetch the first 3 rows of the 'act' DataFrame, where there is a null value under the 'd_name' column.
act[act['d_name'].isna()].head(3)

cds_code  c_code  cd_code  s_code r_type s_name d_name       c_name  \
1064  3.400000e+13    34.0      0.0     0.0      C   None    NaN   Sacramento   
1065  3.900000e+13    39.0      0.0     0.0      C   None    NaN  San Joaquin   
1066  1.500000e+13    15.0      0.0     0.0      C   None    NaN         Kern   

      enroll_12  num_tst_takr avg_scr_read avg_scr_eng avg_scr_math  \
1064    19540.0        3065.0           21          20           21   
1065    11778.0        1348.0           22          21           21   
1066    14229.0        1257.0           21          20           20   

     avg_scr_sci num_ge_21 pct_ge_21     year  
1064          21      1526     49.79  2018-19  
1065          21       693     51.41  2018-19  
1066          20       544     43.28  2018-19

It seems that there could be no School and District Names, but County Names. 

In [29]:
# Fetch the value counts of the 's_name' column of the 'act' DataFrame, where there is a null value under the 'd_name' column.
act[act['d_name'].isna()]['s_name'].value_counts()

None    58
Name: s_name, dtype: int64

All the School Names are 'None'.

In [30]:
# Fetch the value counts of the 'c_name' column of the 'act' DataFrame, where there is a null value under the 'd_name' column.
act[act['d_name'].isna()]['c_name'].value_counts()

Sacramento         1
Placer             1
Tuolumne           1
Napa               1
Butte              1
Yuba               1
Solano             1
Plumas             1
San Bernardino     1
Lassen             1
Alpine             1
Colusa             1
Del Norte          1
Alameda            1
Mariposa           1
San Joaquin        1
Tulare             1
Tehama             1
Fresno             1
San Luis Obispo    1
Monterey           1
Yolo               1
Imperial           1
Siskiyou           1
Santa Clara        1
San Francisco      1
Merced             1
Nevada             1
Modoc              1
Riverside          1
San Diego          1
Amador             1
Kern               1
Shasta             1
Sutter             1
Calaveras          1
Mono               1
Ventura            1
Marin              1
El Dorado          1
Mendocino          1
Glenn              1
Sierra             1
San Mateo          1
San Benito         1
Sonoma             1
Kings              1
Trinity      

In [31]:
# Count the number of values in the list above.
act[act['d_name'].isna()]['c_name'].value_counts().sum()

58

No School Names, no District Names, 58 County Names. It is safe to say that these are the rows for the aggregated data of each county, for the 58 counties of California, and our checks are consistent with this. 

### Replacing the null values in the `'d_name'` column  

We will simply fill the null values with the string 'None'.

In [32]:
act['d_name'].fillna(value='None', inplace=True)

### Replacing the null values in the rest of the columns

In [33]:
act.isna().sum()

cds_code          0
c_code            0
cd_code           0
s_code            0
r_type            0
s_name            0
d_name            0
c_name            0
enroll_12         0
num_tst_takr      0
avg_scr_read    356
avg_scr_eng     356
avg_scr_math    356
avg_scr_sci     356
num_ge_21       356
pct_ge_21       356
year              0
dtype: int64

At this point we have a number of columns, all having 356 values. Let's check the `'avg_scr_read'` (the average score for Reading) column first. 

In [34]:
act[act['avg_scr_read'].isna()].head(3)

cds_code  c_code    cd_code     s_code r_type  \
0  3.366993e+13    33.0  3366993.0   129882.0      S   
1  1.964212e+13    19.0  1964212.0  1995596.0      S   
2  1.563776e+13    15.0  1563776.0  1530377.0      S   

                            s_name                 d_name       c_name  \
0  21st Century Learning Institute       Beaumont Unified    Riverside   
1      ABC Secondary (Alternative)            ABC Unified  Los Angeles   
2      Abraham Lincoln Alternative  Southern Kern Unified         Kern   

   enroll_12  num_tst_takr avg_scr_read avg_scr_eng avg_scr_math avg_scr_sci  \
0       18.0           0.0          NaN         NaN          NaN         NaN   
1       58.0           0.0          NaN         NaN          NaN         NaN   
2       18.0           0.0          NaN         NaN          NaN         NaN   

  num_ge_21 pct_ge_21     year  
0       NaN       NaN  2018-19  
1       NaN       NaN  2018-19  
2       NaN       NaN  2018-19

It seems that these are the columns with no test takers (`'num_tst_takr'`). 

In [35]:
# For the DataFrame rows fetched where it is a null value in the 'avg_scr_read' column, is it equal to 
# the DataFrame rows fetched where the values in the 'num_tst_takr' column are '0.0'?
act[act['avg_scr_read'].isna()].equals(act[act['num_tst_takr'] == 0.0])

True

It is true that the entries (rows) with no test takers are the ones having NaN values under `'avg_scr_read'`. 

Let's apply this test for the rest of the columns, up till the `'pct_ge_21'` column.

In [36]:
# Defining a checker function for the above operation.
def na_has_no_test_takers(col_names, test_takers_col_name, df):
    for col_name in col_names:
        print(col_name + ':', df[df[col_name].isna()].equals(df[df[test_takers_col_name] == 0.0]))

In [37]:
act_list = ['avg_scr_read',
            'avg_scr_eng',
            'avg_scr_math',
            'avg_scr_sci',
            'num_ge_21',
            'pct_ge_21']

na_has_no_test_takers(act_list, 'num_tst_takr', act)

avg_scr_read: True
avg_scr_eng: True
avg_scr_math: True
avg_scr_sci: True
num_ge_21: True
pct_ge_21: True


The data has been validated. There are no anomalous entries for that part of the data. We will fill the null values with the value 'None', since there are no test takers. However, this means there will be multiple datatypes in the same column. We will be excluding the non-test takers before doing summary statistics. 

In [38]:
# Replaces all the null values with 'None' in the columns of interest.
for column in act_list: 
    act[column].fillna(value='None', inplace=True)

In [39]:
# Gets the total number of null values in the entire DataFrame
act.isna().sum().sum()

0

In [40]:
act.isna().sum()

cds_code        0
c_code          0
cd_code         0
s_code          0
r_type          0
s_name          0
d_name          0
c_name          0
enroll_12       0
num_tst_takr    0
avg_scr_read    0
avg_scr_eng     0
avg_scr_math    0
avg_scr_sci     0
num_ge_21       0
pct_ge_21       0
year            0
dtype: int64

We have removed all the null values from the `'act'` DataFrame. 

### Preparing the columns for typecasting 

As we have established, the mixture of datatypes in a column can lead to issues. Another area of concern would be typecasting. 

In [41]:
act.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2309 entries, 0 to 2308
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cds_code      2309 non-null   float64
 1   c_code        2309 non-null   float64
 2   cd_code       2309 non-null   float64
 3   s_code        2309 non-null   float64
 4   r_type        2309 non-null   object 
 5   s_name        2309 non-null   object 
 6   d_name        2309 non-null   object 
 7   c_name        2309 non-null   object 
 8   enroll_12     2309 non-null   float64
 9   num_tst_takr  2309 non-null   float64
 10  avg_scr_read  2309 non-null   object 
 11  avg_scr_eng   2309 non-null   object 
 12  avg_scr_math  2309 non-null   object 
 13  avg_scr_sci   2309 non-null   object 
 14  num_ge_21     2309 non-null   object 
 15  pct_ge_21     2309 non-null   object 
 16  year          2309 non-null   object 
dtypes: float64(6), object(11)
memory usage: 324.7+ KB


There are numeric columns (`'avg_scr_read'` to `'pct_ge_21'`) that should be numeric, but are of object datatype. 

### Dropping the rows with no test-takers

There will be a loss of some information, such as that for enrollment numbers of schools without test-takers. Thus, we can save the DataFrame at this stage. 

In [42]:
act.to_csv('../data/act_2019_ca_before_drop.csv', index=False)

In [43]:
# Reassign the 'act' DataFrame to one including only rows where 'num_tst_takr' is not the float '0.0'
act = act[act['num_tst_takr'] != 0.0]

In [44]:
len(act)

1953

We have removed the rows with no test takers. Originally, it was 2310 rows. The 'None' values have been cleared. 

### Dropping the rows with asterisks 

However, a number of rows still have asterisk values. We know from the data source that this is to preserve anonymity for cases where the number of test takers is very low (14 and below). This is for the columns `'avg_scr_read'` up till `'pct_ge_21'`.

In [45]:
# Fetch the rows of the 'act' DataFrame where 'avg_scr_read' is an asterisk symbol.
act[act['avg_scr_read'] == '*'].head()

cds_code  c_code    cd_code     s_code r_type  \
7   1.964246e+13    19.0  1964246.0   126003.0      S   
8   3.868478e+13    38.0  3868478.0   119958.0      S   
9   3.675077e+13    36.0  3675077.0  3631207.0      S   
10  1.964733e+13    19.0  1964733.0   126573.0      S   
11  3.667736e+13    36.0  3667736.0   116723.0      S   

                                               s_name  \
7                    Academies of the Antelope Valley   
8                          Academy (The)- SF @McAteer   
9                     Academy for Academic Excellence   
10  Academy for Multilingual Arts and Science at M...   
11                 Academy of Careers and Exploration   

                        d_name          c_name  enroll_12  num_tst_takr  \
7   Antelope Valley Union High     Los Angeles       14.0           1.0   
8        San Francisco Unified   San Francisco       84.0           9.0   
9         Apple Valley Unified  San Bernardino       97.0          13.0   
10         Los Angeles Unified     Los Angeles      141.0          10.0   
11        Helendale Elementary  San Bernardino       63.0           5.0   

   avg_scr_read avg_scr_eng avg_scr_math avg_scr_sci num_ge_21 pct_ge_21  \
7             *           *            *           *         *         *   
8             *           *            *           *         *         *   
9             *           *            *           *         *         *   
10            *           *            *           *         *         *   
11            *           *            *           *         *         *   

       year  
7   2018-19  
8   2018-19  
9   2018-19  
10  2018-19  
11  2018-19

Since there is no score information, we will drop these too before doing summary statistics. 

In [46]:
# Reassign the act DataFrame to one including only rows where 'avg_scr_read' is not '*'.
act = act[act['avg_scr_read'] != '*']

Let's check if we have also removed the asterisks from the rest of the columns:

In [47]:
# Get the total number of values in the selected DataFrame columns that match the string '*'
act[['avg_scr_eng','avg_scr_math', 'avg_scr_sci', 'num_ge_21', 'pct_ge_21']].eq('*').sum().sum()

0

Yes, we have removed the asterisks.

In [48]:
len(act)

1421

Previously, it was 1953 rows. Hence, there were no provided values for a few hundred rows. 

### Typecasting columns to numerical form

The datatypes of each column is as follows:

In [49]:
act.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1421 entries, 3 to 2306
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cds_code      1421 non-null   float64
 1   c_code        1421 non-null   float64
 2   cd_code       1421 non-null   float64
 3   s_code        1421 non-null   float64
 4   r_type        1421 non-null   object 
 5   s_name        1421 non-null   object 
 6   d_name        1421 non-null   object 
 7   c_name        1421 non-null   object 
 8   enroll_12     1421 non-null   float64
 9   num_tst_takr  1421 non-null   float64
 10  avg_scr_read  1421 non-null   object 
 11  avg_scr_eng   1421 non-null   object 
 12  avg_scr_math  1421 non-null   object 
 13  avg_scr_sci   1421 non-null   object 
 14  num_ge_21     1421 non-null   object 
 15  pct_ge_21     1421 non-null   object 
 16  year          1421 non-null   object 
dtypes: float64(6), object(11)
memory usage: 199.8+ KB


The numeric columns from the previous part still have to be formally typecasted. 

In [50]:
# Defining a function to convert a given list of columns to a numeric datatype.
def numeric_converter(col_names, df):
    for col_name in col_names:
        df[col_name] = df[col_name].apply(pd.to_numeric)

In [51]:
numeric_converter(act_list, act)

In [52]:
act.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1421 entries, 3 to 2306
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cds_code      1421 non-null   float64
 1   c_code        1421 non-null   float64
 2   cd_code       1421 non-null   float64
 3   s_code        1421 non-null   float64
 4   r_type        1421 non-null   object 
 5   s_name        1421 non-null   object 
 6   d_name        1421 non-null   object 
 7   c_name        1421 non-null   object 
 8   enroll_12     1421 non-null   float64
 9   num_tst_takr  1421 non-null   float64
 10  avg_scr_read  1421 non-null   int64  
 11  avg_scr_eng   1421 non-null   int64  
 12  avg_scr_math  1421 non-null   int64  
 13  avg_scr_sci   1421 non-null   int64  
 14  num_ge_21     1421 non-null   int64  
 15  pct_ge_21     1421 non-null   float64
 16  year          1421 non-null   object 
dtypes: float64(7), int64(5), object(5)
memory usage: 199.8+ KB


The typecasting was successful. 

### Checking the numerical columns for anomalies

We will now check each column to see if it is reasonable. 

The first 4 columns are just codes, but rendered in floats. We would likely prefer them in integer format.

In [53]:
act.head(3)

cds_code  c_code    cd_code     s_code r_type  \
3  4.369666e+13    43.0  4369666.0  4333795.0      S   
4  1.964733e+13    19.0  1964733.0  1935121.0      S   
5  1.976968e+13    19.0  1976968.0   109926.0      S   

                        s_name                         d_name       c_name  \
3         Abraham Lincoln High               San Jose Unified  Santa Clara   
4  Abraham Lincoln Senior High            Los Angeles Unified  Los Angeles   
5      Academia Avance Charter  SBE - Academia Avance Charter  Los Angeles   

   enroll_12  num_tst_takr  avg_scr_read  avg_scr_eng  avg_scr_math  \
3      463.0          53.0            23           22            22   
4      226.0          19.0            21           20            23   
5       65.0          33.0            15           15            17   

   avg_scr_sci  num_ge_21  pct_ge_21     year  
3           23         34      64.15  2018-19  
4           22         11      57.89  2018-19  
5           16          3       9.09  2018-19

In [54]:
# Typecasting the columns to integer format 
act[['cds_code', 'c_code', 'cd_code', 's_code']] = act[['cds_code', 'c_code', 'cd_code', 's_code']].astype('int64')

In [55]:
act.head(3)

cds_code  c_code  cd_code   s_code r_type  \
3  43696664333795      43  4369666  4333795      S   
4  19647331935121      19  1964733  1935121      S   
5  19769680109926      19  1976968   109926      S   

                        s_name                         d_name       c_name  \
3         Abraham Lincoln High               San Jose Unified  Santa Clara   
4  Abraham Lincoln Senior High            Los Angeles Unified  Los Angeles   
5      Academia Avance Charter  SBE - Academia Avance Charter  Los Angeles   

   enroll_12  num_tst_takr  avg_scr_read  avg_scr_eng  avg_scr_math  \
3      463.0          53.0            23           22            22   
4      226.0          19.0            21           20            23   
5       65.0          33.0            15           15            17   

   avg_scr_sci  num_ge_21  pct_ge_21     year  
3           23         34      64.15  2018-19  
4           22         11      57.89  2018-19  
5           16          3       9.09  2018-19

The format is now more readable. 

### Retaining only school-level data

The data also has district, county and state-level data. Such aggregated data would skew the summary statistics and further statistical analysis unless they are excluded. Hence, they will be removed. First, the DataFrame will be saved to back up the information and the cleaning at this stage. 

In [56]:
act.to_csv('../data/act_2019_ca_before_2nd_drop.csv', index=False)

Dropping the values:

In [57]:
# Reassign the 'act' DataFrame to one including only rows where 'r_type' is 'S'
act = act[act['r_type'] == 'S']

In [58]:
# Get the number of rows in the ACT DataFrame. 
len(act)

1016

There are now 1255 rows. Previously, it was 1421 rows. 

We will print some summary statistics for the next few columns. The summary statistics will skip the text-based columns. 

In [59]:
act.describe()

cds_code       c_code       cd_code        s_code    enroll_12  \
count  1.016000e+03  1016.000000  1.016000e+03  1.016000e+03  1016.000000   
mean   2.869969e+13    28.032480  2.869968e+06  2.273475e+06   371.831693   
std    1.340141e+13    13.375514  1.340140e+06  1.725433e+06   206.213129   
min    1.316170e+12     1.000000  1.316170e+05  1.000650e+05     0.000000   
25%    1.964733e+13    19.000000  1.964733e+06  1.344065e+05   173.000000   
50%    3.066522e+13    30.000000  3.066522e+06  1.939156e+06   382.000000   
75%    3.768346e+13    37.000000  3.768346e+06  3.730068e+06   524.000000   
max    5.872770e+13    58.000000  5.872769e+06  6.120158e+06  1135.000000   

       num_tst_takr  avg_scr_read  avg_scr_eng  avg_scr_math  avg_scr_sci  \
count   1016.000000   1016.000000  1016.000000   1016.000000  1016.000000   
mean      79.144685     21.870079    21.017717     21.270669    21.189961   
std       66.574849      3.823038     4.209031      3.499833     3.335189   
min       15.000000     12.000000    10.000000     14.000000    12.000000   
25%       33.000000     19.000000    18.000000     18.000000    19.000000   
50%       58.000000     22.000000    21.000000     21.000000    21.000000   
75%      105.000000     25.000000    24.000000     24.000000    24.000000   
max      512.000000     32.000000    32.000000     32.000000    31.000000   

         num_ge_21    pct_ge_21  
count  1016.000000  1016.000000  
mean     43.673228    51.306850  
std      50.111339    26.534076  
min       0.000000     0.000000  
25%      12.000000    28.570000  
50%      24.000000    52.860000  
75%      55.000000    74.700000  
max     392.000000   100.000000

Given the information we have found in the Background and Outside Research sections, the summary statistics of the data look reasonably correct. The maximums, minimums and averages are within their logical ranges/bounds. 

### Saving the data

In [60]:
act.to_csv('../data/act_2019_ca_cleaned.csv', index=False)

This is the end of the data cleaning portion for the ACT dataset. The data cleaning for the SAT dataset will be continued in the next notebook.